### PR TITLE
chore: Add real world topology benchmark

### LINKED ISSUE #5946: Add topology level benchmarks that do more event manipulation
While discussing #5374 it was identified that we don't really have high level benchmarks that would have shown the regression shown by the test-harness..

We'd like to add criterion benches that would show the same regressions as these are lighter weight, run in CI, and are easier to iterate against then the test-harness tests.

The added benches should show a regression for the changes in #5374

### PR DESCRIPTION
**Pull request by MOZGIII**
Closes #5946.

To do:
- [x] complete the implementation
- [ ] ~test against the #5374 to ensure the performance difference is captured~ we'll merge this one, and rebase #5374 over to have it run at CI

### CODE REVIEW

File: benches/topology.rs [hunk 1]
```diff
@@ -434,10 +434,357 @@ fn benchmark_complex(c: &mut Criterion) {
...
+                topology,
+                output_lines_company_api,
+                output_lines_company_admin,
+                output_lines_company_media_proxy,
+                output_lines_company_unfurler,
+                output_lines_audit,
+            )| {
+                rt.block_on(async move {
+                    // Generate the inputs.
+                    let lines = [
```
Comment (jszwedko) [inline benches/topology.rs:744]: Should we move the input generation up into the setup block?

**Review by jszwedko — APPROVED** [review]
Just one note about the input generation, but otherwise this looks good to me!

I'm curious to see if it shows the regression on the lookup PR.

File: benches/topology.rs (hunk 1 above)
Comment (MOZGIII) [reply benches/topology.rs]: Done! The effect is not significant though - below the noise threshold.

**Comment by MOZGIII** [comment]
> I'm curious to see if it shows the regression on the lookup PR.

I propose that we merge this into `master`, and then into the lookup PR from there - so that we run the diffs via CI.

**Comment by jszwedko** [comment]
> > I'm curious to see if it shows the regression on the lookup PR.
> 
> I propose that we merge this into `master`, and then into the lookup PR from there - so that we run the diffs via CI.

Sounds good!

**Pull request merged**